tests/integration/v0/test_integration.py

```diff
@@ -2,19 +2,243 @@
+IP_ROUTER_PROVIDER_APP_NAME = "ip-router-provider"
+IP_ROUTER_REQUIRER_APP_NAME = "ip-router-requirer"
 
 
-@pytest.mark.abort_on_fail
-async def test_build_and_deploy(ops_test: OpsTest):
-    pass
+def copy_lib_content() -> None:
+    os.makedirs(f"{REQUIRER_CHARM_DIR}/{LIB_DIR}", exist_ok=True)
+    os.makedirs(f"{PROVIDER_CHARM_DIR}/{LIB_DIR}", exist_ok=True)
+    shutil.copyfile(src=f"{LIB_DIR}/{LIB_NAME}", dst=f"{REQUIRER_CHARM_DIR}/{LIB_DIR}/{LIB_NAME}")
+    shutil.copyfile(src=f"{LIB_DIR}/{LIB_NAME}", dst=f"{PROVIDER_CHARM_DIR}/{LIB_DIR}/{LIB_NAME}")
+
+
+async def validate_routing_table(unit, expected_rt) -> None:
+    action = await unit.run_action(action_name="get-routing-table")
+    action_output = await ops_test.model.get_action_output(  # noqa
+        action_uuid=action.entity_id, wait=60
+    )
+    assert json.loads(action_output["msg"]) == expected_rt
+
+
+class TestIntegration:
+    requirer_charm = None
+    provider_charm = None
+
+    @pytest.mark.abort_on_fail
+    async def test_given_charms_packed_when_deploy_charm_then_status_is_blocked(self, ops_test):
+        copy_lib_content()
+        TestIntegration.requirer_charm = await ops_test.build_charm(f"{REQUIRER_CHARM_DIR}/")
+        TestIntegration.provider_charm = await ops_test.build_charm(f"{PROVIDER_CHARM_DIR}/")
+        await ops_test.model.deploy(
+            TestIntegration.provider_charm,
+            application_name=IP_ROUTER_PROVIDER_APP_NAME,
+            series="jammy",
+        )
+        await ops_test.model.deploy(
+            TestIntegration.requirer_charm,
+            application_name=f"{IP_ROUTER_REQUIRER_APP_NAME}-a",
+            series="jammy",
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[f"{IP_ROUTER_REQUIRER_APP_NAME}-a", IP_ROUTER_PROVIDER_APP_NAME],
+            status="blocked",
+            timeout=1000,
+        )
+
+    @pytest.mark.abort_on_fail
+    async def test_given_charms_deployed_when_relate_then_status_is_active(self, ops_test):
+        await ops_test.model.add_relation(
+            relation1=f"{IP_ROUTER_REQUIRER_APP_NAME}-a",
+            relation2=IP_ROUTER_PROVIDER_APP_NAME,
+        )
+
+        await ops_test.model.wait_for_idle(
+            apps=[f"{IP_ROUTER_REQUIRER_APP_NAME}-a", IP_ROUTER_PROVIDER_APP_NAME],
+            status="active",
+            timeout=1000,
+        )
+
+        # Check that the router has the correct app name in the routing table
+        provider_unit = ops_test.model.units[f"{IP_ROUTER_PROVIDER_APP_NAME}/0"]
+
+        action = await provider_unit.run_action(action_name="get-routing-table")
+        action_output = await ops_test.model.get_action_output(
+            action_uuid=action.entity_id, wait=60
+        )
+        assert json.loads(action_output["msg"]) == {f"{IP_ROUTER_REQUIRER_APP_NAME}-a": []}
+
+    @pytest.mark.abort_on_fail
+    async def test_given_network_request_provider_implements_and_requirer_sees(self, ops_test):
```

Comment:
Need arrange the name by specifying "when" and "then"

Reply:
Done